dot/parachain/prospective-parachains/messages.go

```diff
@@ -0,0 +1,292 @@
+package prospective_parachains
+
+import (
+	"fmt"
+
+	parachaintypes "github.com/ChainSafe/gossamer/dot/parachain/types"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/pkg/scale"
+)
+
+type ProspectiveParachainMessageValues interface {
+	IntroduceSecondedCandidate | CandidateBacked | GetBackableCandidates | GetHypotheticalMembership |
+		GetMinimumRelayParents
+}
+
+// ProspectiveParachainsMessage Messages sent to the Prospective Parachains subsystem.
+type ProspectiveParachainsMessage struct {
+	inner any
+}
+
+func setProspectiveParachainMessage[Value ProspectiveParachainMessageValues](ppm *ProspectiveParachainsMessage,
+	value Value) {
+	ppm.inner = value
+}
+
+func (ppm *ProspectiveParachainsMessage) SetValue(value any) (err error) {
+	switch value := value.(type) {
+	case IntroduceSecondedCandidate:
+		setProspectiveParachainMessage(ppm, value)
+		return
+
+	case CandidateBacked:
+		setProspectiveParachainMessage(ppm, value)
+		return
+
+	case GetBackableCandidates:
+		setProspectiveParachainMessage(ppm, value)
+		return
+
+	case GetHypotheticalMembership:
+		setProspectiveParachainMessage(ppm, value)
+		return
+
+	case GetMinimumRelayParents:
+		setProspectiveParachainMessage(ppm, value)
+		return
+
+	default:
+		return fmt.Errorf("unsupported type")
+	}
+}
+
+func (ppm ProspectiveParachainsMessage) IndexValue() (index uint, value any, err error) {
+	switch ppm.inner.(type) {
+	case IntroduceSecondedCandidate:
+		return 0, ppm.inner, nil
+
+	case CandidateBacked:
+		return 1, ppm.inner, nil
+
+	case GetBackableCandidates:
+		return 2, ppm.inner, nil
+
+	case GetHypotheticalMembership:
+		return 3, ppm.inner, nil
+
+	case GetMinimumRelayParents:
+		return 4, ppm.inner, nil
+	}
+
+	return 0, nil, scale.ErrUnsupportedVaryingDataTypeValue
+}
+
+func (ppm ProspectiveParachainsMessage) Value() (value any, err error) {
+	_, value, err = ppm.IndexValue()
+	return
+}
+
+func (ppm ProspectiveParachainsMessage) ValueAt(index uint) (value any, err error) {
+	switch index {
+	case 0:
+		return *new(IntroduceSecondedCandidate), nil
+
+	case 1:
+		return *new(CandidateBacked), nil
+
+	case 2:
+		return *new(GetBackableCandidates), nil
+
+	case 3:
+		return *new(GetHypotheticalMembership), nil
+
+	case 4:
+		return *new(GetMinimumRelayParents), nil
+
+	}
+	return nil, scale.ErrUnknownVaryingDataTypeValue
+}
```

Comment:
We don't need to implement this big code. We implement it when we want to match encode/decode of type with Rust to participate in and share data in the network.

I have provided another way to implement this type of enum In one of the comments below.

Reply:
Updated to simplify.